common/src/main/java/io/netty/util/concurrent/DefaultPromise.java

```diff
@@ -431,24 +432,27 @@ public void run() {
     }
 
     private void notifyListeners0() {
-        Object listeners;
-        while (!notifyingListeners) {
+        Object listeners = null;
+        for (;;) {
             synchronized (this) {
-                if (this.listeners == null) {
+                // Only proceed if there are listeners to notify and we are not already notifying listeners.
```

Comment:
This looks correct but is a bit hard to follow. Maybe special-case the first iteration? More lines of code but seems more readable and less prone to error.

```java
Object listeners;
synchronized (this) {
  if (notifyingListeners || this.listeners == null) {
    return;
  }
  notifyingListeners = true;
  listeners = this.listeners;
  this.listeners = null;
}
for (;;) {
  blahNotifyListener0Stuff();
  synchronized (this) {
    if (this.listeners == null) {
      notifyingListeners = false;
      return;
    }
    listeners = this.listeners;
    this.listeners = null;
  }
}
```